src/equations/ideal_mhd_multiion_2d.jl

```diff
@@ -0,0 +1,888 @@
+# By default, Julia/LLVM does not use fused multiply-add operations (FMAs).
+# Since these FMAs can increase the performance of many numerical algorithms,
+# we need to opt-in explicitly.
+# See https://ranocha.de/blog/Optimizing_EC_Trixi for further details.
+@muladd begin
+
+
+@doc raw"""
+    IdealMhdMultiIonEquations2D
+
+The ideal compressible multi-ion MHD equations in two space dimensions.
+"""
+mutable struct IdealMhdMultiIonEquations2D{NVARS, NCOMP, RealT<:Real} <: AbstractIdealMhdMultiIonEquations{2, NVARS, NCOMP}
+  gammas            ::SVector{NCOMP, RealT} # Heat capacity ratios
+  charge_to_mass    ::SVector{NCOMP, RealT} # Charge to mass ratios
+
+  function IdealMhdMultiIonEquations2D{NVARS, NCOMP, RealT}(gammas       ::SVector{NCOMP, RealT},
+                                                                     charge_to_mass::SVector{NCOMP, RealT}) where {NVARS, NCOMP, RealT<:Real}
+
+    NCOMP >= 1 || throw(DimensionMismatch("`gammas` and `charge_to_mass` have to be filled with at least one value"))
+
+    new(gammas, charge_to_mass)
+  end
+end
+
+function IdealMhdMultiIonEquations2D(; gammas, charge_to_mass)
+
+  _gammas         = promote(gammas...)
+  _charge_to_mass = promote(charge_to_mass...)
+  RealT           = promote_type(eltype(_gammas), eltype(_charge_to_mass))
+
+  NVARS = length(_gammas) * 5 + 3
+  NCOMP = length(_gammas)
+
+  __gammas        = SVector(map(RealT, _gammas))
+  __charge_to_mass = SVector(map(RealT, _charge_to_mass))
+
+  return IdealMhdMultiIonEquations2D{NVARS, NCOMP, RealT}(__gammas, __charge_to_mass)
+end
+
+@inline Base.real(::IdealMhdMultiIonEquations2D{NVARS, NCOMP, RealT}) where {NVARS, NCOMP, RealT} = RealT
+
+have_nonconservative_terms(::IdealMhdMultiIonEquations2D) = True()
+
+function varnames(::typeof(cons2cons), equations::IdealMhdMultiIonEquations2D)
+
+  cons  = ("B1", "B2", "B3")
+  for i in eachcomponent(equations)
+    cons = (cons..., tuple("rho_" * string(i),"rho_v1_" * string(i), "rho_v2_" * string(i), "rho_v3_" * string(i), "rho_e_" * string(i))...)
```

Comment:
This is probably not type stable, thus I wonder if this needs to be fixed. It would probably be safer to do so.